libusb/os/windows_winusb.c

```diff
@@ -2017,6 +2017,157 @@ static int winusb_get_device_list(struct libusb_context *ctx, struct discovered_
 	return r;
 }
 
+struct string_descriptor_s {
+	UCHAR bLength;
+	UCHAR bDescriptorType;
+	uint8_t bString[252];  // UTF-16LE, must be even length
+};
+
+struct string_descriptor_req_s {
+	USB_DESCRIPTOR_REQUEST req;
+	struct string_descriptor_s desc;
+};
+
+static int usbi_utf16le_to_utf8(uint8_t const* src, int src_length, char* dst, int dst_length) {
+	int count = 0;
+	uint16_t codepoint;
+	int overflow = 0;
+
+	for (int k = 0; k < src_length; k += 2) {
+		codepoint = *src++;
+		codepoint |= ((uint16_t)*src++) << 8;
```

Comment:
I imagine there's a -Wcast-align warning here. Better to make `src` `uint16_t*` I think.

Reply:
I don't see any errors / warnings on the Visual Studio build, and I don't think there should be.  We are dereferencing the uint8_t pointer to get a uint8_t value that is then widened to uint16_t.  This is supposed to be standard unpacking code to support little endian regardless of platform endianness.  Do you see something that I am missing?

Reply:
```c
#include <stdint.h>

int main (void)
{
	uint8_t buffer[252];
	uint16_t* ptr = (uint16_t*)buffer;
	
	return 0;
}
```

```
% clang -fsyntax-only -Weverything test.c
test.c:6:18: warning: cast from 'uint8_t *' (aka 'unsigned char *') to 'uint16_t *' (aka 'unsigned short *') increases required alignment from 1 to 2 [-Wcast-align]
        uint16_t* ptr = (uint16_t*)buffer;
                        ^~~~~~~~~~~~~~~~~
```

Reply:
Right, but that is not what the code is doing.  It is doing:
```
(uint16_t) *buffer
```
which recasts the value (dereferenced pointer), not the pointer itself.

Reply:
Oh indeed, I hallucinated the `*` inside the cast. :)

The cast doesn't do much, since the result will be promoted to `unsigned int` anyway, maybe that's why I saw the `*` inside.

Reply:
We do need the cast since we are left shifting to the "upper" byte.  Without the cast, the left-shifted value would be zero.

I will leave the code as is, at least pending any additional feedback.

Reply:
>We do need the cast since we are left shifting to the "upper" byte. Without the cast, the left-shifted value would be zero.

C promotes things to (unsigned) int a lot. So the uint8_t gets implicitly promoted to `unsigned int`. The cast is harmless, but even it's result will get implicitly promoted to `unsigned int`.

See: https://godbolt.org/z/hGrWT84rn

I'm not proposing to remove the cast, it helps readability.